src/Npgsql/NpgsqlCommand.cs

```diff
@@ -1038,47 +1037,58 @@ public override Task<object> ExecuteScalarAsync(CancellationToken cancellationTo
         #region Execute Reader
 
         /// <summary>
-        /// Executes the CommandText against the Connection, and returns an DbDataReader.
+        /// Executes the command text against the connection.
         /// </summary>
-        /// <remarks>
-        /// Unlike the ADO.NET method which it replaces, this method returns a Npgsql-specific
-        /// DataReader.
-        /// </remarks>
-        /// <returns>A DbDataReader object.</returns>
-        public new NpgsqlDataReader ExecuteReader() => (NpgsqlDataReader) base.ExecuteReader();
+        /// <returns>A task representing the operation.</returns>
+        protected override DbDataReader ExecuteDbDataReader(CommandBehavior behavior)
+            => ExecuteReader(behavior);
 
         /// <summary>
-        /// Executes the CommandText against the Connection, and returns an DbDataReader using one
-        /// of the CommandBehavior values.
+        /// Executes the command text against the connection.
         /// </summary>
-        /// <remarks>
-        /// Unlike the ADO.NET method which it replaces, this method returns a Npgsql-specific
-        /// DataReader.
-        /// </remarks>
-        /// <returns>A DbDataReader object.</returns>
-        public new NpgsqlDataReader ExecuteReader(CommandBehavior behavior) => (NpgsqlDataReader) base.ExecuteReader(behavior);
+        /// <param name="behavior">An instance of <see cref="CommandBehavior"/>.</param>
+        /// <param name="cancellationToken">The token to monitor for cancellation requests.</param>
+        /// <returns>A task representing the asynchronous operation.</returns>
+        protected override async Task<DbDataReader> ExecuteDbDataReaderAsync(CommandBehavior behavior, CancellationToken cancellationToken)
+            => await ExecuteReaderAsync(behavior, cancellationToken);
```

Comment:
Call AsTask on the ValueTask returned by ExecuteReaderAsync and remove async from method.

Reply:
Types are not covariant. `ExecuteReaderAsync` returns `Task<NpgsqlReader>` now, but `ExecuteDbDataReaderAsync` should return `Task<DbDataReader>`.

Reply:
OK.

I think it still may be worth checking if the ValueTask compeleted synchronously, and if so, extract the value and return a completed task with it - this way we avoid the state machine overhead. If the ValueTask hasn't completed we can call into an async local method (slow path).

Reply:
Sorry, please disregard the above - ExecuteReaderAsync always completes asynchronously so this would have no value whatsoever.